File: Ios/BarkCloud/BarkCloud/Features/Files/UI/CloudBrowserViewModel.swift
import Foundation
import Observation
import BarkCloudKit

struct CloudBrowserUiState {
    var directoryID: String       // "" = корень
    var title: String
    var crumbs: [PathCrumb] = []
    var subdirs: [CloudDirectory] = []
    var files: [CloudFileEntry] = []
    var isLoading = true
    var isUploading = false
    var snackbar: String?
    /// URL созданной публичной ссылки → системный Share Sheet.
    var pendingShareURL: ShareableURL?

    // MARK: Мультивыбор
    var isSelecting = false
    /// Выбранные файлы и папки — по их `id` (entry_id / directory_id).
    var selectedFiles: Set<String> = []
    var selectedDirs: Set<String> = []
    /// Идёт батч-операция (перемещение/удаление) — показываем прогресс вместо кнопок.
    var isProcessing = false
    var processDone = 0
    var processTotal = 0

    var isEmpty: Bool { subdirs.isEmpty && files.isEmpty }
    var selectedCount: Int { selectedFiles.count + selectedDirs.count }
    var hasSelection: Bool { selectedCount > 0 }
}

@MainActor
@Observable
final class CloudBrowserViewModel {
    var state: CloudBrowserUiState

    /// Очередь отложенного удаления — внизу экрана показывается snackbar с
    /// отсчётом, до выполнения запрос на сервер не уходит.
    let pendingDelete = PendingDelete()

    private let cloud: CloudRepository
    private var didLoad = false

    init(directoryID: String, title: String, cloud: CloudRepository) {
        self.cloud = cloud
        self.state = CloudBrowserUiState(directoryID: directoryID, title: title)
    }

    var isRoot: Bool { state.directoryID.isEmpty }

    func loadIfNeeded() async {
        guard !didLoad else { return }
        didLoad = true
        await reload()
    }

    /// Перезагрузка содержимого папки. `showSpinner` поднимает полноэкранный
    /// `ProgressView` — нужно при программных обновлениях (создание/переименование/
    /// удаление/перемещение). При pull-to-refresh передаём `false`: иначе экран
    /// свернул бы `List` (носитель `.refreshable`) в `ProgressView`, SwiftUI отменил
    /// бы задачу обновления и gRPC-запрос упал бы с «the transport threw an
    /// unexpected error». Спиннер потягивания рисует сам `.refreshable`.
    func reload(showSpinner: Bool = true) async {
        // Если в очереди есть отложенное удаление — выполняем его до перезагрузки,
        // иначе сервер вернёт нам обратно файл, который пользователь только что
        // визуально убрал.
        await pendingDelete.flushIfAny()
        if showSpinner { state.isLoading = true }
        do {
            let listing = try await cloud.listDirectory(state.directoryID)
            state.subdirs = listing.subdirs
            state.files = listing.files
            if !isRoot {
                state.crumbs = (try? await cloud.path(directoryID: state.directoryID)) ?? []
            }
        } catch {
            state.snackbar = domainErrorMessage(error)
        }
        state.isLoading = false
    }

    func createFolder(name: String) async {
        let trimmed = name.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return }
        do {
            _ = try await cloud.createDirectory(parentID: state.directoryID, name: trimmed)
            await reload()
        } catch {
            state.snackbar = domainErrorMessage(error)
        }
    }

    func renameDirectory(_ dir: CloudDirectory, newName: String) async {
        do { try await cloud.renameDirectory(dir.id, newName: newName); await reload() }
        catch { state.snackbar = domainErrorMessage(error) }
    }

    func renameFile(_ entry: CloudFileEntry, newName: String) async {
        do { try await cloud.renameFileEntry(entry.id, newName: newName); await reload() }
        catch { state.snackbar = domainErrorMessage(error) }
    }

    /// Оптимистичное удаление папки: сразу убираем из сетки и кладём в очередь
    /// отложенного удаления — реальный gRPC-запрос уйдёт, когда snackbar отсчитает
    /// свои 5 секунд (или пользователь поставит другое удаление в очередь).
    func deleteDirectory(_ dir: CloudDirectory) {
        guard let index = state.subdirs.firstIndex(where: { $0.id == dir.id }) else { return }
        state.subdirs.remove(at: index)
        pendingDelete.schedule(
            label: dir.name,
            action: { [weak self, cloud] in
                do { try await cloud.deleteDirectory(dir.id) }
                catch {
                    // Сервер не дал удалить — возвращаем папку на место и сообщаем.
                    self?.state.snackbar = domainErrorMessage(error)
                    await self?.reload(showSpinner: false)
                }
            },
            onUndo: { [weak self] in
                guard let self else { return }
                let position = min(index, state.subdirs.count)
                state.subdirs.insert(dir, at: position)
            }
        )
    }

    /// Оптимистичное удаление файла (см. `deleteDirectory`).
    func deleteFile(_ entry: CloudFileEntry) {
        guard let index = state.files.firstIndex(where: { $0.id == entry.id }) else { return }
        state.files.remove(at: index)
        pendingDelete.schedule(
            label: entry.name,
            action: { [weak self, cloud] in
                do {
                    try await cloud.deleteFileEntry(entry.id)
                    await DeviceCopyCleaner.deleteDeviceCopy(forCloudFileID: entry.fileID)
                } catch {
                    self?.state.snackbar = domainErrorMessage(error)
                    await self?.reload(showSpinner: false)
                }
            },
            onUndo: { [weak self] in
                guard let self else { return }
                let position = min(index, state.files.count)
                state.files.insert(entry, at: position)
            }
        )
    }

    func moveDirectory(_ dir: CloudDirectory, toDirectory targetID: String) async {
        do { try await cloud.moveDirectory(dir.id, newParentID: targetID); await reload() }
        catch { state.snackbar = domainErrorMessage(error) }
    }

    func moveFile(_ entry: CloudFileEntry, toDirectory targetID: String) async {
        do { try await cloud.moveFileEntry(entry.id, newDirectoryID: targetID); await reload() }
        catch { state.snackbar = domainErrorMessage(error) }
    }

    /// Поставить файлы в фоновую очередь (`BackgroundUploadCoordinator`). UI не
    /// ждёт завершения — загрузка переживёт сворачивание/kill приложения. По
    /// событию `onJobCompleted` в `AppEnvironment` файл будет привязан к папке.
    func upload(_ files: [(data: Data, fileName: String)]) async {
        guard !files.isEmpty else { return }
        var anyFailed = false
        for file in files {
            do {
                _ = try await cloud.enqueueBackgroundUpload(
                    data: file.data,
                    fileName: file.fileName,
                    toDirectory: state.directoryID,
                    source: .manual
                )
            } catch {
                anyFailed = true
            }
        }
        if anyFailed { state.snackbar = String(localized: "upload_failed") }
    }

    func snackbarShown() { state.snackbar = nil }

    /// Создать публичную ссылку на файл и открыть системный Share Sheet.
    /// Зеркалит `GalleryViewModel.makePublic` / `MediaGridViewModel.makePublic`.
    func makePublic(_ entry: CloudFileEntry) async {
        do {
            let link = try await cloud.createShare(fileID: entry.fileID, name: entry.name)
            guard let url = link.url else {
                state.snackbar = String(localized: "shared_load_failed")
                return
            }
            state.pendingShareURL = ShareableURL(url: url)
        } catch {
            state.snackbar = domainErrorMessage(error)
        }
    }

    // MARK: - Мультивыбор

    func enterSelection() {
        guard !state.isLoading else { return }
        state.isSelecting = true
    }

    func exitSelection() {
        state.isSelecting = false
        state.selectedFiles = []
        state.selectedDirs = []
    }

    func toggleFile(_ entry: CloudFileEntry) {
        if state.selectedFiles.contains(entry.id) {
            state.selectedFiles.remove(entry.id)
        } else {
            state.selectedFiles.insert(entry.id)
        }
    }

    func toggleDirectory(_ dir: CloudDirectory) {
        if state.selectedDirs.contains(dir.id) {
            state.selectedDirs.remove(dir.id)
        } else {
            state.selectedDirs.insert(dir.id)
        }
    }

    /// Батч-удаление выбранных папок и файлов — последовательно, с прогрессом.
    /// Без undo (для одиночного удаления есть отложенная очередь со snackbar).
    func deleteSelected() async {
        let dirs = state.subdirs.filter { state.selectedDirs.contains($0.id) }
        let files = state.files.filter { state.selectedFiles.contains($0.id) }
        guard !dirs.isEmpty || !files.isEmpty else { return }
        state.isProcessing = true
        state.processTotal = dirs.count + files.count
        state.processDone = 0
        var anyFailed = false
        for dir in dirs {
            do { try await cloud.deleteDirectory(dir.id) } catch { anyFailed = true }
            state.processDone += 1
        }
        var deletedFileIDs: [String] = []
        for entry in files {
            do { try await cloud.deleteFileEntry(entry.id); deletedFileIDs.append(entry.fileID) }
            catch { anyFailed = true }
            state.processDone += 1
        }
        // Убираем копии удалённых файлов и с устройства (один системный диалог).
        await DeviceCopyCleaner.deleteDeviceCopies(forCloudFileIDs: deletedFileIDs)
        finishBatch(anyFailed: anyFailed)
        await reload(showSpinner: false)
    }

    /// Батч-перемещение выбранных папок и файлов в папку `targetID`.
    func moveSelected(toDirectory targetID: String) async {
        let dirs = state.subdirs.filter { state.selectedDirs.contains($0.id) }
        let files = state.files.filter { state.selectedFiles.contains($0.id) }
        guard !dirs.isEmpty || !files.isEmpty else { return }
        state.isProcessing = true
        state.processTotal = dirs.count + files.count
        state.processDone = 0
        var anyFailed = false
        for dir in dirs {
            do { try await cloud.moveDirectory(dir.id, newParentID: targetID) } catch { anyFailed = true }
            state.processDone += 1
        }
        for entry in files {
            do { try await cloud.moveFileEntry(entry.id, newDirectoryID: targetID) } catch { anyFailed = true }
            state.processDone += 1
        }
        finishBatch(anyFailed: anyFailed)
        await reload(showSpinner: false)
    }

    private func finishBatch(anyFailed: Bool) {
        state.isProcessing = false
        state.processTotal = 0
        state.processDone = 0
        exitSelection()
        if anyFailed { state.snackbar = String(localized: "cloud_batch_failed") }
    }
}

File: Android/BarkCloud.Android/app/src/main/java/com/barkfluff/BarkCloud/ui/files/CloudBrowserViewModel.kt
package com.barkfluff.BarkCloud.ui.files

import android.content.Context
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.CreationExtras
import com.barkfluff.BarkCloud.BarkCloudApplication
import com.barkfluff.BarkCloud.data.cloud.CloudDirectory
import com.barkfluff.BarkCloud.data.cloud.CloudFileEntry
import com.barkfluff.BarkCloud.data.cloud.CloudRepository
import com.barkfluff.BarkCloud.data.cloud.PathCrumb
import com.barkfluff.BarkCloud.data.upload.UploadScheduler
import com.barkfluff.BarkCloud.net.queryFileName
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class CloudBrowserUiState(
    val isLoading: Boolean = true,
    val crumbs: List<PathCrumb> = emptyList(),
    val subdirs: List<CloudDirectory> = emptyList(),
    val files: List<CloudFileEntry> = emptyList(),
    val isUploading: Boolean = false,
    val snackbar: String? = null,
) {
    val isEmpty: Boolean get() = !isLoading && subdirs.isEmpty() && files.isEmpty()
}

class CloudBrowserViewModel(
    private val appContext: Context,
    private val cloudRepository: CloudRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(CloudBrowserUiState())
    val state: StateFlow<CloudBrowserUiState> = _state.asStateFlow()

    private var directoryId: String = ""
    private var started = false

    fun start(directoryId: String) {
        if (started) return
        started = true
        this.directoryId = directoryId
        reload()
    }

    fun reload() {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true) }
            try {
                val listing = cloudRepository.listDirectory(directoryId)
                val crumbs = if (directoryId.isEmpty()) emptyList()
                else runCatching { cloudRepository.path(directoryId) }.getOrDefault(emptyList())
                _state.update {
                    it.copy(
                        isLoading = false,
                        subdirs = listing.subdirs,
                        files = listing.files,
                        crumbs = crumbs,
                    )
                }
            } catch (e: Exception) {
                _state.update { it.copy(isLoading = false, snackbar = e.message) }
            }
        }
    }

    fun createFolder(name: String) = mutate { cloudRepository.createDirectory(directoryId, name.trim()) }
    fun renameDirectory(id: String, newName: String) = mutate { cloudRepository.renameDirectory(id, newName.trim()) }
    fun deleteDirectory(id: String) = mutate { cloudRepository.deleteDirectory(id) }
    fun renameFile(entryId: String, newName: String) = mutate { cloudRepository.renameFileEntry(entryId, newName.trim()) }
    fun deleteFile(entryId: String) = mutate { cloudRepository.deleteFileEntry(entryId) }
    fun moveDirectory(id: String, target: String) = mutate { cloudRepository.moveDirectory(id, target) }
    fun moveFile(entryId: String, target: String) = mutate { cloudRepository.moveFileEntry(entryId, target) }

    fun upload(uris: List<Uri>) {
        if (uris.isEmpty()) return
        _state.update { it.copy(isUploading = true) }
        viewModelScope.launch {
            var failures = 0
            uris.forEach { uri ->
                runCatching {
                    (appContext as BarkCloudApplication).uploadQueue.enqueue(uri, queryFileName(appContext, uri), directoryId)
                }
                    .onFailure { failures++ }
            }
            if (failures < uris.size) UploadScheduler.enqueue(appContext)
            _state.update { it.copy(isUploading = false, snackbar = if (failures == 0) UPLOAD_QUEUED else UPLOAD_PARTIAL) }
            reload()
        }
    }

    private fun mutate(action: suspend () -> Unit) {
        viewModelScope.launch {
            try {
                action()
                reload()
            } catch (e: Exception) {
                _state.update { it.copy(snackbar = e.message) }
            }
        }
    }

    fun snackbarShown() = _state.update { it.copy(snackbar = null) }

    companion object {
        private const val UPLOAD_PARTIAL = "Часть файлов не загрузилась"
        private const val UPLOAD_QUEUED = "Загрузка поставлена в очередь"

        fun factory(): ViewModelProvider.Factory = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>, extras: CreationExtras): T {
                val app = extras[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY]
                    as BarkCloudApplication
                return CloudBrowserViewModel(app, app.cloudRepository) as T
            }
        }
    }
}
